### PR TITLE
chore: two happy path system tests for lender side -- supply and redeem

### PR DESCRIPTION
**Pull request by charleenfei**

### CODE REVIEW

File: src/core/test/system/test_utils.sol [hunk 1]
```diff
@@ -90,6 +90,31 @@ contract User is DSTest{
     }
 }
 
+contract WhitelistInvestor is DSTest{
```
Comment (manuelpolzhofer) [inline src/core/test/system/test_utils.sol]:
```suggestion
contract WhitelistInvestor is DSTest{
```
```suggestion
contract WhitelistedInvestor is DSTest{
```

File: src/core/lender/deployer.sol [hunk 2]
```diff
@@ -93,11 +93,18 @@ contract DistributorLike {
...
     function deny(address usr) public;
 }
 
+contract WhitelistOperatorLike {
+    function rely(address usr) public;
+    function deny(address usr) public;
+    function relyInvestor(address usr) public;
+    function denyInvestor(address usr) public;
+}
+
```
Comment (manuelpolzhofer) [inline src/core/lender/deployer.sol]: I want to avoid that.

The idea is to have only an `OperatorFab`. 

You need to cast it in the test to a whitelist Operator

File: src/core/test/system/system.t.sol [hunk 3]
```diff
@@ -321,4 +324,23 @@ contract SystemTest is TestUtils, DSTest {
         borrower.doBorrow(loan, 100);
         assertEq(currency.balanceOf(borrower_), 100);
     }
+
+    function testSupply() public {
```
Comment (manuelpolzhofer) [inline src/core/test/system/system.t.sol]: lets make a own file for lender tests

File: src/core/test/system/test_utils.sol (hunk 1 above)
Comment (manuelpolzhofer) [reply src/core/test/system/test_utils.sol]: I think we don't need to make it DSTest

File: src/core/test/system/test_utils.sol [hunk 4]
```diff
@@ -90,6 +90,31 @@ contract User is DSTest{
...
+
+    constructor(address operator_, address currency_, address token_) public {
+        operator = WhitelistOperator(operator_);
+        currency = ERC20Like(currency_);
+        token = ERC20Like(token_);
+    }
+
+    function doSupply(uint amount) public {
+        address tranche_ = address(operator.tranche());
+        currency.approve(tranche_, uint(-1));
```
Comment (manuelpolzhofer) [inline src/core/test/system/test_utils.sol]:
```suggestion
        currency.approve(tranche_, amount);
```

File: src/core/test/system/test_utils.sol [hunk 5]
```diff
@@ -90,6 +90,31 @@ contract User is DSTest{
...
+        operator.supply(amount);
+    }
+
+    function doRedeem(uint amount) public {
+        address tranche_ = address(operator.tranche());
+        currency.approve(tranche_, uint(-1));
```
Comment (manuelpolzhofer) [inline src/core/test/system/test_utils.sol]:
```suggestion

```
not needed

File: src/core/test/system/test_utils.sol [hunk 6]
```diff
@@ -90,6 +90,31 @@ contract User is DSTest{
...
+        token.approve(tranche_, uint(-1));
```
Comment (manuelpolzhofer) [inline src/core/test/system/test_utils.sol]:
```suggestion
        token.approve(tranche_, amount);
```

File: src/core/test/system/system.t.sol [hunk 7]
```diff
@@ -140,6 +138,11 @@ contract SystemTest is TestUtils, DSTest {
         return currency.balanceOf(address(lenderDeployer.distributor()));
     }
 
+    function supply(uint supply, uint amount) public {
```
Comment (manuelpolzhofer) [inline src/core/test/system/system.t.sol]:
```suggestion
    function supply(uint balance, uint amount) public {
```

The naming is a bit confusing

File: src/core/test/system/system.t.sol [hunk 8]
```diff
@@ -321,4 +324,23 @@ contract SystemTest is TestUtils, DSTest {
         borrower.doBorrow(loan, 100);
         assertEq(currency.balanceOf(borrower_), 100);
     }
+
+    function testSupply() public {
+        supply(100 ether, 10 ether);
```
Comment (manuelpolzhofer) [inline src/core/test/system/system.t.sol]: I think we should use variable and not constant numbers if we have more than one parameter.

```javascript
uint investorBalance;
uint supplyAmount;
supply(investorBalance, supplyAmount);
```

File: src/core/test/system/system.t.sol [hunk 9]
```diff
@@ -321,4 +324,23 @@ contract SystemTest is TestUtils, DSTest {
...
+    function testRedeem() public {
+        supply(100 ether, 10 ether);
+        assertEq(currency.balanceOf(juniorInvestor_), 90 ether);
+        assertEq(lenderDeployer.juniorERC20().balanceOf(juniorInvestor_), 10 ether);
+        // additional liquidity changes the token price
+        // TODO: add code to move funds from shelf to tranche so we can enact the actual repayment scenario
+        currency.mint(address(lenderDeployer.junior()), 100 ether);
+        juniorInvestor.doRedeem(10 ether);
+        assertEq(lenderDeployer.juniorERC20().balanceOf(juniorInvestor_), 0);
+        assertEq(currency.balanceOf(juniorInvestor_), 190 ether);
```
Comment (manuelpolzhofer) [inline src/core/test/system/system.t.sol]: Shouldn't it be `200 ether`

I think the test is not working because you need to change the flag in the distributor to repay.

File: src/core/test/system/system.t.sol [hunk 10]
```diff
@@ -43,18 +44,15 @@ contract SystemTest is TestUtils, DSTest {
...
 
-        // todo replace with investor contract
-        rootAdmin.relyLenderAdmin(address(this));
-
-        // give invest rights to test
-        WhitelistOperator juniorOperator = WhitelistOperator(address(lenderDeployer.juniorOperator()));
-        juniorOperator.relyInvestor(address(this));
-
+        // junior investor
+        rootAdmin.relyLenderAdmin(juniorInvestor_);
```
Comment (manuelpolzhofer) [inline src/core/test/system/system.t.sol]: The investor should be an admin.  The investor should be only allowed to call the tranche.

File: src/core/test/system/root_admin.sol [hunk 11]
```diff
@@ -29,6 +29,7 @@ contract TestRootAdmin is RootAdmin {
     // Needed for System Tests
     function relyLenderAdmin(address usr) public auth {
         lenderDeployer.juniorOperator().rely(usr);
+        lenderDeployer.juniorOperator().relyInvestor(usr);
```
Comment (manuelpolzhofer) [inline src/core/test/system/root_admin.sol]: I think an admin shouldn't be an investor by default

File: src/core/test/system/system.t.sol [hunk 12]
```diff
@@ -43,18 +44,15 @@ contract SystemTest is TestUtils, DSTest {
...
         // give admin access rights to contract
         // root only for this test setup
         rootAdmin.relyBorrowAdmin(admin_);
 
-        // todo replace with investor contract
-        rootAdmin.relyLenderAdmin(address(this));
-
-        // give invest rights to test
-        WhitelistOperator juniorOperator = WhitelistOperator(address(lenderDeployer.juniorOperator()));
-        juniorOperator.relyInvestor(address(this));
```
Comment (manuelpolzhofer) [inline src/core/test/system/system.t.sol]: why did you remove it?

File: src/core/test/system/lender/redeem.t.sol [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.5.12;
+
+import "../system.t.sol";
+import {SwitchableDistributor} from "../../../lender/distributor/switchable.sol";
+
+contract RedeemTest is SystemTest {
```
Comment (manuelpolzhofer) [inline src/core/test/system/lender/redeem.t.sol]:
```suggestion
contract RedeemTest is TestUtils {
```

If we use system tests we execute all system test twice.

But maybe it is a good pattern to move a default `setUp` method into TestUtils

File: src/core/test/system/system.t.sol [hunk 14]
```diff
@@ -101,11 +103,16 @@ contract SystemTest is TestUtils, DSTest {
...
+        // borrow transaction
         borrower.doBorrow(loan, principal);
         checkAfterBorrow(tokenId, principal);
     }
 
+    function supply(uint balance, uint amount) public {
+        currency.mint(juniorInvestor_, balance);
+        juniorInvestor.doSupply(amount);
+    }
+
```
Comment (manuelpolzhofer) [inline src/core/test/system/system.t.sol]: I would move it to testUtils

File: src/core/lender/distributor/base.sol [hunk 15]
```diff
@@ -150,11 +151,13 @@ contract BaseDistributor is Math, DSNote, Auth {
 
         // repay senior always first
         if(address(senior) != address(0)) {
+            currency.approve(address(senior), available);
```
Comment (manuelpolzhofer) [inline src/core/lender/distributor/base.sol]: maybe lets move it to `_repay`

File: src/core/test/system/lender/redeem.t.sol (hunk 13 above)
Comment (charleenfei) [reply src/core/test/system/lender/redeem.t.sol]: after conversation with @ilinzweilin and @manuelpolzhofer we have decided to move this issue to another PR

File: src/core/test/system/lender/redeem.t.sol [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        uint supplyAmount = 10 ether;
+        supply(investorBalance, supplyAmount);
+        assertEq(currency.balanceOf(juniorInvestor_), 90 ether);
+        assertEq(lenderDeployer.juniorERC20().balanceOf(juniorInvestor_), 10 ether);
+
+        SwitchableDistributor switchable = SwitchableDistributor(address(lenderDeployer.distributor()));
+        switchable.file("borrowFromTranches", false);
+        juniorInvestor.doRedeem(10 ether);
+        assertEq(lenderDeployer.juniorERC20().balanceOf(juniorInvestor_), 0);
+        assertEq(currency.balanceOf(juniorInvestor_), 100 ether);
```
Comment (ilinzweilin) [inline src/core/test/system/lender/redeem.t.sol]: same as in the comment below, would be helpful to use the vars directly in the assertions

File: src/core/test/system/lender/redeem.t.sol (hunk 13 above)
Comment (ilinzweilin) [reply src/core/test/system/lender/redeem.t.sol]: yes, I would even create a clean base SystemTest contract that inherits from testUtils but also has some additional methods like setUp, assertPrecondition, assertPostcondition... and let the tests inherit from it.

File: src/core/test/system/users/whitelisted_investor.sol [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity >=0.5.12;
+
+import "ds-test/test.sol";
+import "../../../lender/deployer.sol";
+import "../test_utils.sol";
```
Comment (ilinzweilin) [inline src/core/test/system/users/whitelisted_investor.sol]: maybe you can also add the interfaces you need to the interfaces file in the test folder and import it, then you can remove the test_utils & deployer imports.

File: src/core/test/system/test_utils.sol [hunk 18]
```diff
@@ -72,11 +71,8 @@ contract User is DSTest{
     }
 
     function doRepay(uint loan, uint wad, address usr) public {
-        emit log_named_uint("loan", wad);
```
Comment (ilinzweilin) [inline src/core/test/system/test_utils.sol]: thank you 🙈

File: src/core/test/system/users/whitelisted_investor.sol (hunk 17 above)
Comment (charleenfei) [reply src/core/test/system/users/whitelisted_investor.sol]: good idea, I will add it in the TestUtils Pr
